features/FEATURE_BLE/source/gap/AdvertisingDataBuilder.cpp

```diff
@@ -0,0 +1,469 @@
+    }
+
+    /* Check if data fits */
+    if ((_payload_length + fieldData.size()) <= _buffer.size()) {
+        uint8_t old_data_length = field[0];
+
+        /* get the size of bytes in the payload after the field */
+        size_t remainder_size = _payload_length -
+                                (field - _buffer.data()) - /* length of all data before the field */
+                                (old_data_length + 1) /* length of the old field */;
+
+        /* move data after the field to fit new data */
+        if (remainder_size) {
+            memmove(
+                field + old_data_length + 1 + fieldData.size(),
+                field + old_data_length + 1,
+                remainder_size
+            );
+        }
+
+        /* append new data */
+        memcpy(field + old_data_length + 1, fieldData.data(), fieldData.size());
+
+        /* Increment lengths */
+        field[0] += fieldData.size();
+        _payload_length += fieldData.size();
+
+        return BLE_ERROR_NONE;
+    } else {
+        return BLE_ERROR_BUFFER_OVERFLOW;
+    }
+}
+
+ble_error_t AdvertisingDataBuilder::replaceField(
+    adv_data_type_t advDataType,
+    mbed::Span<const uint8_t> fieldData,
+    uint8_t* field
+) {
+    if (fieldData.size() > 0xFE) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    uint8_t old_data_length = field[0] - 1;
+
+    /* New data has same length, do in-order replacement */
+    if (fieldData.size() == old_data_length) {
+        memcpy(field + 2, fieldData.data(), old_data_length);
+
+        return BLE_ERROR_NONE;
+    } else {
+        /* Check if data fits */
+        if ((_payload_length - old_data_length + fieldData.size()) <= _buffer.size()) {
+            removeField(field);
+
+            /* Add new field */
+            return addField(advDataType, fieldData);
+        } else {
+            return BLE_ERROR_BUFFER_OVERFLOW;
+        }
+    }
+}
+
+ble_error_t AdvertisingDataBuilder::removeField(uint8_t* field)
+{
+    /* stored length + the byte containing length */
+    uint8_t old_field_length = field[0] + 1;
+
+    memmove(field, field + old_field_length, old_field_length);
+
+    _payload_length -= old_field_length;
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t AdvertisingDataBuilder::setUUIDData(
+    mbed::Span<const UUID> data,
+    adv_data_type_t shortType,
+    adv_data_type_t longType
+) {
+    ble_error_t status = BLE_ERROR_NONE;
+
+    /* first count all the bytes we need to store all the UUIDs */
+    size_t size_long = 0;
+    size_t size_short = 0;
+
+    for (size_t i = 0, end = data.size(); i < end; ++i) {
+        if (data[i].shortOrLong() == UUID::UUID_TYPE_SHORT) {
+            size_short++;
+        } else {
+            size_long++;
+        }
+    }
+
+    if ((size_long * 8 > 0xFE) || (size_short * 2 > 0xFE)) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    /* calculate total size including headers for types */
+    size_t total_size = size_long + (!!size_long) * 2 +
```

Comment:
An explanation of exactly why this calculation is structured this way would be most helpful

Reply:
fixed